packages/liveblocks-react/rollup.config.js

```diff
@@ -1,19 +1,108 @@
+ */
+import path from "path";
 import resolve from "@rollup/plugin-node-resolve";
+import babelPlugin from "@rollup/plugin-babel";
+import typescript from "@rollup/plugin-typescript";
+import esbuild from "rollup-plugin-esbuild";
+import dts from "rollup-plugin-dts";
+import { promises } from "fs";
+const createBabelConfig = require("./babel.config");
 
-import pkg from "./package.json";
+const extensions = [".tsx"];
 
-export default {
-  input: "src/index.tsx",
-  output: [
+const external = ["react", "@liveblocks/client"];
+
+function getBabelOptions(targets) {
+  return {
+    ...createBabelConfig({ env: (env) => env === "build" }, targets),
+    extensions,
+    comments: false,
+    babelHelpers: "bundled",
+  };
+}
+
+function getEsbuild(target) {
+  return esbuild({
+    minify: false,
+    target,
+    tsconfig: path.resolve("./tsconfig.build.json"),
+  });
+}
+
+/**
+ * We use @rollup/plugin-typescript to generate typescript definition
+ * but it does not support the declarationOnly option so we delete
+ * js files and bundle d.ts files with rollup-plugin-dts
+ */
+function createDeclarationConfig(input, output) {
+  return [
+    {
+      input,
+      output: {
+        dir: "lib",
+      },
+      external,
+      plugins: [
+        typescript({
+          declaration: true,
+          outDir: `./lib/tmp`, // We need to put it in "lib" because of typescript can't @rollup/plugin-typescript ouput outside tsconfig outDir option
+          tsconfig: "./tsconfig.build.json",
+        }),
+      ],
+    },
     {
-      file: pkg.main,
-      format: "cjs",
-      exports: "named",
-      sourcemap: true,
-      strict: false,
+      input: `./lib/tmp/${output}.d.ts`,
+      output: [{ file: `lib/${output}.d.ts`, format: "es" }],
+      plugins: [
+        dts(),
+        {
+          closeBundle: async () => {
+            await promises.rm(`./lib/tmp`, {
+              recursive: true,
+              force: true,
+            });
+          },
+        },
+      ],
     },
-  ],
-  plugins: [typescript(), resolve()],
-  external: ["react", "@liveblocks/client"],
+  ];
+}
+
+function createCommonJSConfig(input, output) {
+  return {
+    input,
+    output: { file: `lib/${output}`, format: "cjs", exports: "named" },
+    external,
+    plugins: [
+      resolve({ extensions }),
+      babelPlugin(getBabelOptions({ ie: 11 })),
+    ],
+  };
+}
+
+function createESMConfig(input, output) {
+  return {
+    input,
+    external,
+    output: [
+      { file: `lib/${output}.js`, format: "esm" },
+      { file: `lib/${output}.mjs`, format: "esm" },
+    ],
+    plugins: [getEsbuild("node12")],
+  };
+}
+
+export default async () => {
+  await promises.rm(`./lib`, {
+    recursive: true,
+    force: true,
+  });
+
+  return [
+    ...createDeclarationConfig("src/index.tsx", "index"),
+    createCommonJSConfig("src/index.tsx", "index.js"),
+    createESMConfig("src/index.tsx", "esm/index"),
+  ];
```

Comment:
Hmm. Not sure what's going on here without scrutinizing the exact changes here. This makes the `rollup.config.js` diverge a lot from the other 4 packages though, which is a bit unfortunate. I've kept all of the rest 99% the same. One thing I'd love to try is to use `tsup` for bundling our packages. It supposedly ships with great community defaults and best practices for publishing TypeScript projects as ESM/CJS in a near-zero config way, is super fast to build, and is compatible with a proper NPM monorepo setup, which would also be on my wish list (aka bye bye `link-liveblocks.sh`). It would allow us to get rid of all this manual rollup and plugin configuration we are currently juggling.

Reply:
@GuillaumeSalles From just looking at this PR's diff, I don't understand the cause of the Vite bug and/or the fix for it. I would probably be able to dig in by setting up a Vite project myself and diving in from there, but I'll spare myself the effort and time. When you're back from your holiday, I'd love to better understand the cause/fix made here, though.